tools/istio-iptables/pkg/builder/iptables_builder_impl.go

```diff
@@ -170,6 +177,125 @@ func (rb *IptablesRuleBuilder) buildRules(rules []*Rule) [][]string {
+				modifiedParams = append(modifiedParams, "-D")
+			} else if element == "-I" || element == "--insert" {
+				insertIndex = i
+				modifiedParams = append(modifiedParams, "-D")
+			} else {
+				modifiedParams = append(modifiedParams, element)
+			}
+
+			if ((element == "-A" || element == "--append") || (element == "-I" || element == "--insert")) &&
+				i < len(r.params)-1 && strings.HasPrefix(r.params[i+1], "ISTIO_") {
+				skip = true
+			} else if (element == "-j" || element == "--jump") && i < len(r.params)-1 && strings.HasPrefix(r.params[i+1], "ISTIO_") {
+				skip = false // Override previous skip if this is a jump-rule
+			}
+		}
+		if skip {
+			continue
+		}
+
+		output = append(output, &Rule{
+			chain:  r.chain,
+			table:  r.table,
+			params: modifiedParams,
+		})
+	}
+	log.Debugf("Reversed rules to %+v", output)
+	return output
+}
+
+func checkRules(rules []*Rule) []*Rule {
+	output := make([]*Rule, 0)
+	for _, r := range rules {
+		var modifiedParams []string
+		insertIndex := -1
+		for i, element := range r.params {
+			if insertIndex >= 0 && i == insertIndex+2 {
+				continue
+			}
+			if element == "-A" || element == "--append" {
+				modifiedParams = append(modifiedParams, "-C")
+			} else if element == "-I" || element == "--insert" {
+				insertIndex = i
+				modifiedParams = append(modifiedParams, "-C")
+			} else {
+				modifiedParams = append(modifiedParams, element)
+			}
+		}
+		output = append(output, &Rule{
+			chain:  r.chain,
+			table:  r.table,
+			params: modifiedParams,
+		})
+	}
+	log.Debugf("Rules mutated into check-rules %+v", output)
+	return output
+}
+
+func (rb *IptablesRuleBuilder) buildCheckRules(rules []*Rule) [][]string {
+	output := make([][]string, 0)
+	checkRules := checkRules(rules)
+	for _, r := range checkRules {
+		cmd := append([]string{"-t", r.table}, r.params...)
+		output = append(output, cmd)
+	}
+	return output
+}
+
+func (rb *IptablesRuleBuilder) buildCleanupRules(rules []*Rule) [][]string {
+	newRules := make([]*Rule, len(rules))
+	for i := len(rules) - 1; i >= 0; i-- {
+		newRules[len(rules)-1-i] = rules[i]
+	}
+
+	output := make([][]string, 0)
+	reversedRules := reverseRules(newRules)
+	for _, r := range reversedRules {
+		cmd := append([]string{"-t", r.table}, r.params...)
+		output = append(output, cmd)
+	}
+	chainTableLookupSet := sets.New[string]()
+	for _, r := range newRules {
+		chainTable := fmt.Sprintf("%s:%s", r.chain, r.table)
+		// Delete chain if key: `chainTable` isn't present in map
+		if !chainTableLookupSet.Contains(chainTable) {
+			// Don't delete iptables built-in chains
+			if _, present := constants.BuiltInChainsMap[r.chain]; !present {
+				cmd := []string{"-t", r.table, "-F", r.chain}
+				output = append(output, cmd)
+				cmd = []string{"-t", r.table, "-X", r.chain}
+				output = append(output, cmd)
+				chainTableLookupSet.Insert(chainTable)
+			}
+		}
+	}
+	return output
+}
+
+func (rb *IptablesRuleBuilder) buildGuardrails() []*Rule {
```

Comment:
I don't get this guardrails concept. isn't the point of iptables-save and restore to allow doing transactions? Why do we need to do delete+add when we can just reconcile the state in one go under a lock?

Do other implementations do this?

Reply:
The guardrails I believe are in place because `iptables-save` output is _not_ relied upon, and we have to go rule-by-rule with `iptables -C` as per https://github.com/istio/istio/pull/50328#issuecomment-2098281216

So, we have to DIY transactions, basically.